foursight_core/react/api/react_api.py

```diff
@@ -279,13 +282,55 @@ def reactapi_header(self, request: dict, env: str) -> Response:
             data["auth"]["known_envs"] = [known_envs_default]
             data["auth"]["known_envs_actual_count"] = known_envs_actual_count
         data["auth"]["default_env"] = self._envs.get_default_env()
+        test_mode_certificate_simulate_error = read_cookie_bool(request, "test_mode_certificate_simulate_error")
+        if test_mode_certificate_simulate_error:
+            data["portal"]["url"] = None
+        if not data["portal"]["url"]:
+            # Here we did not get a Portal URL from the to app.core.get_portal_url (via _reactapi_header_nocache).
+            # That call ends up ultimately calling the Portal health endpoint (via s3Utils.get_synthetic_env_config
+            # via environment.get_environment_and_bucket_info). So there may have been a problem with the Portal,
+            # e.g. bad SSL certificate; we will get the Portal URL by other means, and from get get its SSL
+            # certificate to help diagnose any problem with that. C4-1017 (April 2023).
+            try:
+                if test_mode_certificate_simulate_error:
+                    raise Exception("test_mode_certificate_simulate_error")
+                portal_url = app.core.get_portal_url(env or default_env, raise_exception=True)
+                data["portal"]["url"] = portal_url
+            except Exception as e:
+                e = str(e)
+                data["portal"]["exception"] = e
+                if "certifi" in e.lower():
+                    data["portal"]["ssl_certificate_error"] = True
+                portal_url = env_utils_get_portal_url(env)
+                data["portal"]["url"] = portal_url
+                data["portal"]["ssl_certificate"] = get_ssl_certificate_info(portal_url)
+                if data["portal"]["ssl_certificate"]:
+                    data["portal"]["ssl_certificate"]["name"] = "Portal"
+                    data["portal"]["ssl_certificate"]["exception"] = e
         data["timestamp"] = convert_utc_datetime_to_useastern_datetime_string(datetime.datetime.utcnow())
+        test_mode_access_key_simulate_error = read_cookie_bool(request, "test_mode_access_key_simulate_error")
+        if auth.get("user_exception"): # or test_mode_access_key_simulate_error:
+            # Since this call to get the Portal access key info can be relatively expensive, we don't want to
+            # do it on every /header API call; so we only call it if, according to the user's authtoken cookie,
+            # an exception was experienced when trying to authorize the user (via envs.get_user_auth_info) on
+            # login, which if so, would indicate that the is likely a problem with the Portal access key.
+            test_mode_access_key_expiration_warning_days = \
+                read_cookie_int(request, "test_mode_access_key_expiration_warning_days")
+            data["portal_access_key"] = get_portal_access_key_info(
+                env,
+                obfuscate=not is_logged_in,
```

Comment:
I reworked this to make it more clear; if logged in we return the Portal access key ID, otherwise (if not logged in) we obfuscate it. (And we never return any part of the Portal access key secret at all).